### ISSUE TITLE
Specifying the shape of the return attention vectors from the decoder

### ISSUE THREAD

**Issue by ItakeLs** [OpenNMT/CTranslate2#1087]
Hello, I've been playing around with this library, and I noticed that when returning the attention vectors in the beam decoder, ```decode()``` it returns only one head which I'm assuming is from the first layer.

Is there any way to modify the number of attention layers and heads that returned from ```decode()```.

**Comment by guillaumekln** [comment]
Currently we get the attention from a [specific layer](https://github.com/OpenNMT/CTranslate2/blob/v3.5.1/src/layers/transformer.cc#L409) and then [reduce the attention](https://github.com/OpenNMT/CTranslate2/blob/v3.5.1/src/layers/transformer.cc#L421) by averaging one or more attention heads.

If you change this part, you should also update the decoding code to deal with the new attention shape (if it is different).

**Comment by ItakeLs** [comment]
I'm specifically using the whisper model, if I am to return the attention weights used in the dim format [layers, heads, tokens, frames], would that require me to make significant overhauls of the decoding code?

**Comment by guillaumekln** [comment]
> would that require me to make significant overhauls of the decoding code

I don't recall every details, so I guess you'll find out.

> Also from my understanding isn't the whisper encoder sending its output to all the decoder blocks / attention layers, hence wouldn't the decoding function already use the necessary attention weights since your getting identical quality to that of the original whisper?

Yes the model is using the full attention weights. But during beam search you have to collect and reorder these attention weights. I'm just not sure the existing decoding code (in `decoding.cc`) works with any attention shape.

> Also from my understanding isn't the whisper encoder sending its output to all the decoder blocks / attention layers, hence wouldn't the decoding function already use the necessary attention weights since your getting identical quality to that of the original whisper?

Yes, this is from `best_of`.

**Comment by ItakeLs** [comment]
In the greedy search you are adding the reduced attention heads of dimensions ```[batch, frames]``` to the result in this snippet:
```
if (attention_step) {
  const auto* attn = attention_step.index<float>({i, 0});
  results[batch_id].attention[0].emplace_back(attn, attn + attention_step.dim(-1));
}
```

I have managed to get the full attention weights in the dimension ```[batch, layers, head, token, frames]``` and made a new variable to store that in the ```decoding results```.

```
std::vector<std::vector<std::vector<float>>> attention;
std::vector<std::vector<std::vector<std::vector<std::vector<float>>>>> full_attention;
```

My question is how can I append my full attention weights to the variable, I see in the earlier snippet, you don't just ```emplace_back``` the attention weights but instead I think adding the memory. (when I print out ```attn``` I get a memory address) and I'm not really sure what ```attn + attention_step.dim(-1)``` is needed for in ```emplace_back```.

would something like this work is there a reason you append the weights that way?
```
const auto* full_attn = full_attention_step.index<float>({i, 0});
results[batch_id].full_attention[0].emplace_back(full_attn);
```

**Comment by ItakeLs** [comment]
Hi, in this snippet you append the reduced attention weights with dim ```[N, frames]``` like this
```
if (attention_step) {
  const auto* attn = attention_step.index<float>({i, 0});
  results[batch_id].attention[0].emplace_back(attn, attn + attention_step.dim(-1));
}
```

I'm wondering why you ```emplace_back``` using the memory address (if that is what you are doing) instead of just appending the vector entirely. I managed to get the full attention weights with dims ```[N, layers, heads, tokens, frames]``` and created a new variable in the decoder results.

```
std::vector<std::vector<std::vector<float>>> attention;
std::vector<std::vector<std::vector<std::vector<std::vector<float>>>>> full_attention;
```

I tried to append them using the same method but the different dimensions affect the result and I'm not sure how to append "vector" with more than 2 dimensions using your method.

```
if (full_attention_step) {
  const auto* full_attn = full_attention_step.index<float>({i, 0, 0, 0, 0});
  results[batch_id].full_attention[0].emplace_back(full_attn, full_attn + full_attention_step.dim(-1));
}
```

My question is how can I append the full attention weights with those dimension.

**Comment by guillaumekln** [comment]
> I'm wondering why you emplace_back using the memory address (if that is what you are doing) instead of just appending the vector entirely.

We only want to append the `i`-th row. This is what this code is doing. We could build an intermediate `std::vector` but it would have the same effect.

> I tried to append them using the same method but the different dimensions affect the result and I'm not sure how to append "vector" with more than 2 dimensions using your method.

Maybe you should try to save a `StorageView` in the results instead of nested `std::vector` which can be difficult to build correctly.

**Comment by ItakeLs** [comment]
Funnily enough, I just managed to figure out how to append the weights, however, the results I'm getting from the attention are not what I expect (trying to use the attention weights to align the word level timestamp). I just want to clarify if this is what you meant by getting the full attention weights

I have a variable that stores the full attention weights
```
StorageView attention_layers;
```

I get the weights from the output of each layer here
```
        (*_layers[l])(layer_in,
                      input_lengths_mask.get(),
                      memory,
                      memory_lengths_mask.get(),
                      cached_self_attn_keys,
                      cached_self_attn_values,
                      cached_attn_keys,
                      cached_attn_values,
                      layer_out,
                      attention,
                      input_padder.get(),
                      memory_padder.get());
        layer_in = std::move(layer_out);
        if(attention){
          append_attention_layer(attention_layers, *attention);
        }
```

Append the attention weights across the layer dimension (normally, its the first dimension, but you use that for the batches)
```
static void append_attention_layer(StorageView& layers,
                                    StorageView& layer) {
  StorageView new_layer;
  new_layer.copy_from(layer);

  if (layers) {
    //concatnate along the 2 axis [batches, heads, layers, frames] i.e. layers 
    const StorageView cur_layers(std::move(layers));
    ops::Concat(2)({&cur_layers, &new_layer}, layers);
  } else {
    layers = std::move(new_layer);
  }
}
```

then I return the full attention weights as a separate variable, that way I don't need to modify the core functionality of ```decoding .cc``` to keep the attention weights.

**Comment by guillaumekln** [comment]
> I have an additional question, are the attention weights the equivalent of the cross attention weights you'd get from the whisper decoder blocks e.g.

No, it seems they are not the same. Whisper returns the attention weights [before the softmax](https://github.com/openai/whisper/blob/53807677fe8bc892fe0fff970e8b24ddd3848596/whisper/model.py#L98-L101), but in our implementation we return the attention weights after the softmax.

**Comment by ItakeLs** [comment]
Sorry for the ignorance, but is there any way for me to get the weights without the softmax (like whisper). Because don't these snippets basically do the same thing
```
qk = qk.float()

w = F.softmax(qk, dim=-1).to(q.dtype)
return (w @ v).permute(0, 2, 1, 3).flatten(start_dim=2), qk.detach()
```

```
StorageView attn(values.dtype(), values.device());
ops::SoftMax()(output, values_lengths, attn);

const ops::MatMul values_matmul;
values_matmul(attn, values, output);
if (relative_position_values)
  add_relative_representations(attn,
                                *relative_positions,
                                *relative_position_values,
                                values_matmul,
                                output);

if (attention) {
  if (beam_size == 1)
    *attention = std::move(attn);
  else {
    transpose_op(attn, *attention);
    attention->reshape({-1, attn.dim(1), 1, attn.dim(-1)});
  }
}
```

**Comment by ItakeLs** [comment]
Do you have other idea on what might be causing a difference, I am assuming the weights should be identical to whisper, when ```beam_search=None```, ```best_of=1``` and ```temperature=0``` and there is no temperature fallback.

**Comment by guillaumekln** [comment]
What's the difference you are observing?

**Comment by ItakeLs** [comment]
Forget what I am saying, the weights aren't exactly the same but it produces the same functionality, in fact maybe they are the same but in different orders. As long as the functionality is the same it is fine. 

Thank you for your help regarding returning the full attention weights, I really appreciate the assistance. Once I finish testing and improve performance I can send a pull request if you think the feature is relevant and won't cause bloat.

**Comment by KMC07** [comment]
> Maybe you should try to save a `StorageView` in the results instead of nested `std::vector` which can be difficult to build correctly.

Just wanted to ask, when you mentioned saving the vector in a ```StorageView```, how would you append a ```StorageView``` into the first index of a ```StorageView```, and when you return the results will the ```StorageView``` automatically be converted into a nested ```std::vector```.

**Comment by ItakeLs** [comment]
> Maybe you should try to save a `StorageView` in the results instead of nested `std::vector` which can be difficult to build correctly.

Just wanted to ask, when you mentioned saving the vector in a ```StorageView```, how would you append a ```StorageView``` into the first index of a ```StorageView```, and when you return the results will the ```StorageView``` automatically be converted into a nested ```std::vector```.

**Comment by guillaumekln** [comment]
> Just wanted to ask, when you mentioned saving the nested vector in a StorageView, Is It possible to return nested std::vectors and StorageView directly as NumPy matrix without having to copy values.

Yes, search for `StorageView` in the `python` directory and you'll find classes and functions that convert arrays from or to Numpy.

> Additionally, is there a way to access a StorageView  index and still have it be a StorageView , similar to a numpy matrix[0] or tensor matrix[0], because I assume something like StorageView.index<StorageView>({i, :, :, :}) or StorageView[i] wouldn't work.

Once you indexed a dimension, you can create another `StorageView` from the returned pointer. For example using this constructor:

```cpp
    // Create from a buffer (no copy).
    template <typename T>
    StorageView(Shape shape, T* data, Device device = Device::CPU);
```

You just need to resolve the shape correctly.

**Comment by ItakeLs** [comment]
Thank you for the resposne, for some reason when I perform ```ops``` functions such as ```ops::split``` or ```ops::mean``` after using the program crashes.

This is the way I am creating a new storageview from the point
```
auto* full_attn = attention_layers.index<float>({0, 0, _alignment_layer, 0}); //[batch, heads, layers, tokens]
Shape layer_shape = std::vector<dim_t>{attention_layers.dim(0), attention_layers.dim(1), 1, attention_layers.dim(3)};
StorageView attention_layer(layer_shape, full_attn, Device::CPU);
*attention = reduce_multi_head_attention(attention_layers, _alignment_heads);
```

Additionally, I am trying to return the full cross-attention weights efficiently, do you have any suggestions on how one might return them efficiently?

**Comment by sphinxrave** [comment]
Slightly unrelated question: can I access the attention layer weights (read or write) from the Python Whisper interface? Or is this only available if I call the C++ stuff?

**Comment by ItakeLs** [comment]
> Slightly unrelated question: can I access the attention layer weights (read or write) from the Python Whisper interface? Or is this only available if I call the C++ stuff?

You would need to modify some of the c++ and the python to get the attention weight, however, it would only return the weights from the last layer, and the mean of the attention heads.

Getting the full cross-attention weights efficiently would require a lot more modification of the c++ codebase along with the python codebase.
